feat: upgrade Jython runtime from 2.5 to 2.7

tommy-gilligan opened the pull request:
Currently Nodel runs on Jython 2.5.4-rc1.  This is a bit limiting when you try to use 3rd party code in your recipe.

These changes make it possible to use Jython 2.7 instead (expanding what is possible in your recipes).

Leaving it as something to opt-in to for now: you can use Jython 2.5.4-rc1 or Jython 2.7 with these changes in place.
This means there is a bit of branching to do the right thing depending on which version of Jython you are trying to use.  This is maybe not such a huge problem going forward because it doesn't look like there'll be a newer version of Jython any time soon.

tommy-gilligan committed:
Fix invalid cast (PyNone -> PyFunction) in Jython 2.7

tommy-gilligan committed:
Import nodetoolkit from sys when it is there

tommy-gilligan committed:
Use __code__ with Jython 2.7 and func_code with Jython 2.5

justparking commented:
Hey @tomgilligan, nice one. I look forward to checking out what you've done and building the code.

It was a very very long time ago when I think I attempted something similar as a side project. I'm trying to remember exactly what the issue was, but there was one sticking point related to ... ummm I think, callbacks (function references) and them being implemented with Weak referencing in that version of Jython.

It would manifest itself as callbacks basically just going silent over a long period of time.

Anyway, it'd be good for you to run with it an see how long term tests go.

Oh yes, one other thing I noticed was it took MUCH longer to instantiate interpreter instances and deal with higher quantities of nodes.

So test for that too i.e. node count > 50, and see how it goes verses 2.5.4.

(@scroix FYI)

justparking commented:
Hey @tomgilligan, I was expecting more affected files than just the **PyNode.java** file.

I thought this branch would be where Jython 2.7 itself is swapped in and gradle files would be affected, etc.

Is any of that work available too?

tommy-gilligan committed:
Default to Jython 2.7.2

scroix commented:
I gave the binary a whirl. It looks like the toolkit is unavailable.

```
07-18 17:50:05.64 Traceback (most recent call last):
07-18 17:50:05.64   File "C:\Users\scroix\sandbox\ap\nodel\nodel-jython2.7\nodel-jyhost\build\distributions\standalone\.\nodes\test\script.py", line 15, in <module>
07-18 17:50:05.64     param_disabled = Parameter({'desc': 'Disables this node', 'schema': {'type': 'boolean'}})
07-18 17:50:05.64 NameError: name 'Parameter' is not defined
07-18 17:50:05.64 
07-18 17:50:05.64 (Python and Node script(s) loaded with errors (took 395ms); calling 'main' if present...)
07-18 17:50:05.64 (no 'main' to call)
07-18 17:50:35.12 Running Python 2.7.2 👍
```

It looks like it is _there_ e.g. `from sys import nodetoolkit` executes, but there is some silent error and all API functions are unavailable, e.g. `NameError: name 'Parameter' is not defined`.

You can manually import it, but the "injection" seems to of broken.
```
from sys import nodetoolkit
nodetoolkit.systemClockInMillis()
> 123456789
```
As opposed to standard toolkit usage.
```
system_clock()
> 123456789
```

scroix committed:
Update to Jython-2.7.3

(cherry picked from commit 25d6926c6e2251b8dd733209200546774b4f748e)

scroix committed:
Remove Jython-2.5 scenario

(cherry picked from commit 05b2eb4e44b515336fe2d1767caed06e6e9cc422)

scroix committed:
Update sys.path with nodetoolkit whilst injecting

(cherry picked from commit c5876c5b6123d4f579e6823d70dc1f656b9915c5)

scroix commented:
There's a fix for the injection of the toolkit sitting in (https://github.com/tomgilligan/nodel/pull/1) which I've been testing and noticed our first of presumably many issues.

In Jython 2.5, if you tried to access a missing key in a java.util.LinkedHashMap object, it would return None. However, in Jython 2.7, accessing a missing key in a LinkedHashMap object will raise a KeyError exception, just like in a regular Python dictionary.

```python
from java.util import LinkedHashMap

m = LinkedHashMap()
m.put("a", 1)

print(m.get("b"))
```
> In Jython 2.5, this would print None, while in Jython 2.7 it would raise a KeyError exception with the message "KeyError: 'b'".

This behaviour occurs in the [Calendar ](https://github.com/museumsvictoria/nodel-recipes/tree/master/Calendar)recipe as it [parses members](https://github.com/museumsvictoria/nodel-recipes/blob/a59a3bba1cee97e75d3b8076d6d67f1493fd0491/Calendar/script.py#L218) (`traverse(memberInfo)`), and there might be others we'll need to work through.

For example, it appears that In Jython 2.5, the string representation of a LinkedHashMap object used the {key=value} format, while in Jython 2.7, the string representation uses the {'key': 'value'} format.

justparking commented:
Hey @scroix , that dictionary behaviour is one I've thought about for a while and would be nice to address. But as you say, it's a definite possible breaking change for existing recipes that have coding not expecting `myDict['missingKey'] ...` to throw an exception.

But regarding 2.7 vs 2.5, have you done any benchmarks regarding load and compile time for the same set of nodes?

I started doing that but then got side-tracked.

Loading / compiling times aside and a few class behaviour issues, have you noticed anything else?

scroix committed:
Merge branch 'master' into jython-2.7-compatibility

scroix committed:
Fix Jython path init and function resolution

Restores sys.path for nodetoolkit and safely resolves function code.

(cherry picked from commit 0eb34bab3b68484392381156cfc326d05e6e8ef6)

scroix committed:
Add JyHost integration test

- Adds JyHostIntegrationTest to verify script loading and event flow

- Adds JUnit 4 dependency to nodel-jyhost

(cherry picked from commit e61e27be74734642e353eea2a763c02e3c681779)

scroix committed:
fix: use PyUnicode for sys.path to resolve nested node loading issues in Jython 2.7

(cherry picked from commit 02bd979a96610a489782e8dcb05665f577bfdf16)

scroix committed:
perf: disable python.import.site to improve Jython 2.7 startup and script load performance

(cherry picked from commit c75f6ea9595e8ee83441602621b57219d4a2eb29)

scroix committed:
perf: add UTF-8 console encoding and refine Jython 2.7 optimizations

(cherry picked from commit c3f5e25d5eb86f49fb815da05958e8148807eaa0)

scroix committed:
build: update Jython to version 2.7.4

scroix commented:
I've patched this up a bit and have it working locally.

Also upgraded the embedded Jython to [2.7.4](https://central.sonatype.com/artifact/org.python/jython-standalone). The idea, is that this is still a reasonable modernisation effort to think about...

@justparking Initially, things were quite slow (~2.5x) but disabling the standard site-packages import appears to have helped quite a bit. It's still slower, but, more like ~1.5x

scroix committed:
Consolidate interpreter working directory and sys.path setup in injectToolkit

The setup in applyConfig0 mutated a system state the thread-local
interpreter never consults; the injectToolkit copy is the effective one
(removing it breaks script and local module loading). Keep a single,
commented location to prevent the duplication from being 'cleaned up'
in the wrong direction.

scroix committed:
Simplify function code resolution to direct __code__ field access

PyFunction.__code__ is a public typed field in Jython 2.7 and the
func_code fallback became dead code when Jython 2.5 support was
dropped from this branch.

scroix committed:
Tidy lifecycle guards and remove unused PyString import

Perform each _globals lookup once, drop the null-check made redundant
by instanceof, and fix guard-body indentation.

scroix committed:
Extend integration test with lifecycle and local module coverage

The Emitter node now defines main, @before_main, @after_main and
@at_cleanup functions and imports a sibling helper module, asserting
via the console that each ran. @at_cleanup is exercised by touching
script.py and waiting for the interpreter recycle. These cover the
sys.path handling and lifecycle guard paths changed for Jython 2.7.

scroix committed:
Add Jython 2.7 migration notes to README

scroix committed:
Apply Jython property defaults without overriding operator settings

python.import.site and python.console.encoding become defaults rather
than forced values, so a -D system property from the operator wins.

scroix committed:
fix: preserve Jython state while resolving master conflicts
